### PR TITLE
#232 Exception on uploading when no file specified

### PR DESCRIPTION
**Pull request by dmzaytsev**
For #232 
- add the zero-length check for all implementations of `com.nerodesk.om.Doc`
- add new unit tests
- disable upload button if no file specified

### CODE REVIEW

**Comment by k0r** [comment]
@yegor256 good to merge?

**Comment by karato** [comment]
@dmzaytsev Many thanks for the PR, let me find a reviewer for it

**Comment by karato** [comment]
@longtimeago review it

File: src/main/java/com/nerodesk/om/aws/AwsDoc.java [hunk 1]
```diff
@@ -120,6 +120,11 @@ public void read(@NotNull final OutputStream output) throws IOException {
     @Override
     public void write(final InputStream input, final long size)
         throws IOException {
+        if (size == 0) {
```
Comment (longtimeago) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java]: @dmzaytsev Maybe, it would be better to have a decorator `SafeDoc` to check the size and avoid code duplication?

File: src/main/xsl/docs.xsl [hunk 2]
```diff
@@ -48,8 +48,9 @@
                 <xsl:text>My Documents</xsl:text>
             </h2>
             <form method="post" action="{links/link[@rel='upload']/@href}"
-                enctype="multipart/form-data">
-                <input name="file" type="file"/>
+                enctype="multipart/form-data"
+                onSubmit="if(document.getElementById('fileinput').value == '') return false;">
```
Comment (longtimeago) [inline src/main/xsl/docs.xsl]: @dmzaytsev You could try to use 

```
onSubmit="if(this.value == '') return false;"
```

and remove id from the input

File: src/test/java/com/nerodesk/om/aws/AwsDocTest.java [hunk 3]
```diff
@@ -192,6 +192,20 @@ public Void answer(final InvocationOnMock invocation) {
     }
 
     /**
+     * AwsDoc throws exception when the file has zero length.
+     * @throws IOException In case of error
+     */
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsException() throws IOException {
```
Comment (longtimeago) [inline src/test/java/com/nerodesk/om/aws/AwsDocTest.java]: @dmzaytsev this test should be for `SafeDoc`

File: src/test/java/com/nerodesk/om/mock/MkDocTest.java [hunk 4]
```diff
@@ -162,4 +162,17 @@ public void shortenUrl() throws IOException {
             Matchers.equalTo(file.toURI().toURL().toString())
         );
     }
+
+    /**
+     * MkDoc throws exception when the file has zero length.
+     * @throws IOException In case of error
+     */
+    @Test(expected = IllegalArgumentException.class)
+    public void throwsException() throws IOException {
```
Comment (longtimeago) [inline src/test/java/com/nerodesk/om/mock/MkDocTest.java]: @dmzaytsev this would be not needed in case of `SafeDoc`

File: src/main/xsl/docs.xsl (hunk 2 above)
Comment (longtimeago) [inline src/main/xsl/docs.xsl]: @dmzaytsev I would also trim the value before check

File: src/test/java/com/nerodesk/takes/doc/TkSetVisibilityTest.java [hunk 5]
```diff
@@ -75,7 +75,7 @@ public void returnsResponse() throws Exception {
         final String name = "returnsResponse.txt";
         // @checkstyle MultipleStringLiterals (1 line)
         base.user("urn:test:1").docs().doc(name)
-            .write(new ByteArrayInputStream(new byte[0]), 0);
+            .write(new ByteArrayInputStream("X".getBytes()), 1);
```
Comment (longtimeago) [inline src/test/java/com/nerodesk/takes/doc/TkSetVisibilityTest.java:78]: @dmzaytsev why did you change this?

**Comment by longtimeago** [comment]
@dmzaytsev please, see several comments above

File: src/main/java/com/nerodesk/om/aws/AwsDoc.java (hunk 1 above)
Comment (dmzaytsev) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java]: @longtimeago Let's try to imagine a situation when we need to store a zero size Doc ... I can't think of any. If we will use the decorator SafeDoc we will leave the possibility wrong use of the Doc. I don't like a decorator idea in this case, if we don't allow Null parameter in a method we don't create a decorator for this, right? We check it inside the method.

Even if we still need to store zero-length files then for each implementation is done in its own way. E.g. `File.setLastModified()` for MkDoc and some other way for AwsDoc.

File: src/test/java/com/nerodesk/takes/doc/TkSetVisibilityTest.java (hunk 5 above)
Comment (dmzaytsev) [inline src/test/java/com/nerodesk/takes/doc/TkSetVisibilityTest.java:78]: @longtimeago MkDos doesn't allow zero size length thats why I changed this

File: src/main/java/com/nerodesk/om/aws/AwsDoc.java (hunk 1 above)
Comment (longtimeago) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java]: @dmzaytsev We already have may examples of such decorations in our code. For example, SmallDoc.
Or in XHT project. This is what Yegor expected to see. This is how Open/Closed principle works (http://en.wikipedia.org/wiki/Open/closed_principle)
As far as the check is completely the same in both places we should use decoration.

For null parameter we use annotation, so it's not the code duplication.

File: src/test/java/com/nerodesk/takes/doc/TkSetVisibilityTest.java (hunk 5 above)
Comment (longtimeago) [inline src/test/java/com/nerodesk/takes/doc/TkSetVisibilityTest.java:78]: @dmzaytsev Got it

File: src/main/java/com/nerodesk/om/aws/AwsDoc.java (hunk 1 above)
Comment (dmzaytsev) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java]: @longtimeago from the article above

> The idea was that once completed, the implementation of a class could only be modified to correct errors; new or changed features would require that a different class be created.

I think `SmallDoc` is the new feature, a `Doc` instance can be use with or without this.
but zero length checking is not a new feature. This bug is that it not checked before...

never mind I will implement `SafeDoc`

Comment (longtimeago) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java]: @dmzaytsev as you said before, theoretically, we might decide to save doc with zero length in some cases.
This can be controlled by adding SafeDoc decorator or not

Comment (dmzaytsev) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java]: @longtimeago  if we decide to save doc with zero length we should remove SafeDoc and implement the zero-length file storage...
right now if we remove SafeDoc, the Doc instance will be broken on the zero length file

File: src/main/xsl/docs.xsl (hunk 2 above)
Comment (dmzaytsev) [inline src/main/xsl/docs.xsl]: @longtimeago `onSubmit="if(this.value== '') return false;"`
doesn't work... at least in Chrome

Comment (longtimeago) [inline src/main/xsl/docs.xsl]: @dmzaytsev ok, let's leave it as is

**Comment by dmzaytsev** [comment]
@longtimeago fixed
take a look please

File: src/test/java/com/nerodesk/om/SafeDocTest.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            new ByteArrayInputStream(bytes), bytes.length
+        );
+    }
+
+    /**
+     * SafeDoc can shorten a URL.
+     * @throws IOException In case of error.
+     */
+    @Test
+    public void shortenUrl() throws IOException {
```
Comment (longtimeago) [inline src/test/java/com/nerodesk/om/SafeDocTest.java]: @dmzaytsev This test is not related to SafeDoc functionality at all. Please, move it to more appropriate place

File: src/test/java/com/nerodesk/om/SmallDocTest.java [hunk 7]
```diff
@@ -87,7 +87,7 @@ public void failsToWriteBigDoc() throws IOException {
         );
         final String content = StringUtils.repeat('b', 101);
         final byte[] bytes = content.getBytes(StandardCharsets.UTF_8);
-        new SmallDoc(new MkDoc(file, "", ""), Tv.HUNDRED).write(
+        new SafeDoc(new SmallDoc(new MkDoc(file, "", ""), Tv.HUNDRED)).write(
```
Comment (longtimeago) [inline src/test/java/com/nerodesk/om/SmallDocTest.java]: @dmzaytsev why do you need to decor here? The test should fail because of big file, not a zero size

**Comment by longtimeago** [comment]
@dmzaytsev please, see 2 more remarks. Also, I didn't found where you actually plug SafeDoc?

File: src/test/java/com/nerodesk/om/SmallDocTest.java (hunk 7 above)
Comment (dmzaytsev) [inline src/test/java/com/nerodesk/om/SmallDocTest.java]: @longtimeago my mistalke
It should be placed into  another class

**Comment by dmzaytsev** [comment]
@longtimeago done

**Comment by longtimeago** [comment]
@dmzaytsev thank you!

**Comment by longtimeago** [comment]
@rultor merge please

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by karato** [comment]
@longtimeago Thank you, I have added **19 mins** to your account in payment/transaction "`58167455`", time consumed: 20 hours and 2 mins

you're getting extra minutes here (c=4)

+19 to your rating, your total score is [+4233](http://www.netbout.com/b/35024?open=rating)

**Comment by karato** [comment]
@rultor please deploy
